### PR TITLE
Add Gamma sampler

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR is an attempt to implement Gamma sampler based on [this reference](https://dl.acm.org/citation.cfm?doid=358407.358414). The current version is slow so I very appreciate any help from JAX team to address this performance issue.

TODO:
- [x] Address the performance
- [x] Using lax api instead of numpy.

### CODE REVIEW

**Comment by mattjj** [comment]
I haven't looked at the code or thought about the problem, but I wanted to throw out an idea: it can be useful to look at algorithms that are specialized to the GPU because they are often focused on vectorization, and vectorization will be salient both for performance and for expressibility in lax and numpy. If there's a gamma sampling routine in CUDA or a related library, or as a TF kernel, I'd be curious how it's done there.

@slinderman do you know how gammas are sampled, especially on vector machines?

**Comment by mattjj** [comment]
Re: your second todo item, it's okay to use the numpy api instead of the lax one. The lax one just has slightly lower overheads, and fewer accidental casts to f64, because of its simpler broadcasting / dtype promotion logic. But convenience should come first!

**Comment by fehiepsi** [comment]
Thank you @mattjj ! Marsaglia and Tsang's method, which I used, is popular for Gammar sampling (numpy/pytorch uses this method). In CUDA, it seems that [cuRAND](https://docs.nvidia.com/cuda/curand/device-api-overview.html) provides an API to support sampling in parallel. And if I understand correctly, then it is quite similar to what `vmap` does.

**Comment by fehiepsi** [comment]
Actually, in CPU, the jit version is 10x slower than numpy and 4.5x slower than pytorch. It is not bad to me given than the implementation is purely in python. The comparison can be found in [my gist](https://gist.github.com/fehiepsi/58af0bb9a5773decd133f429758556ce).

**Comment by mattjj** [comment]
Well, we think `@jit` can be _faster_ than C. But XLA's while_loop is a known performance cliff.

I will also tag @hawkinsp because he knows how to make XLA code fast. In the worst case, we can make primitives that translate into CPU / GPU library calls as appropriate, like we have done with LAPACK / cuSOLVER.

**Comment by fehiepsi** [comment]
@mattjj Previous implementation requires a while_loop inside a while_loop (kind of an if-else inside an if-else). It turns out that computations of outer loop are quite cheap (only involves a few mul operators) quite the inner loop is expensive (log, uniform sampling, normal sampling). So I join two loops (by moving the inner body_fn outside and merging two cond_fn). Now, this implementation is only 2x slower than pytorch and 3x slower than numpy. I personally satisfy with this performance so I have removed draft state.

You are right that the lax one just has slightly lower overheads and their jaxpr are quite similar. :)

**Comment by slinderman** [comment]
Marsaglia and Tsang (@fehiepsi's reference above) is the standard method for generating gammas.  It's a rejection sampling algorithm with a very high acceptance rate for most parameter regimes.  Rejection samplers do not seem particularly well suited to GPU's since each thread may execute a variable number of iterations, and if one thread rejects then the whole warp waits. I don't have any better suggestions right now.  (I'm assuming the inverse CDF would be even more expensive on a GPU and that integer shape restrictions are out.)  

I wrote a [blog post](https://lips.cs.princeton.edu/a-parallel-gamma-sampling-implementation/) about this many years ago.  With some googling, I found this much [better post](http://cas.ee.ic.ac.uk/people/dt10/research/rngs-gpu-rejection.html).  It suggests some useful performance enhancements.

**Comment by fehiepsi** [comment]
Thanks a lot @slinderman for your suggestions! Warp-based parallelism is neat to me. If I understand it correctly then instead of the sequential while_loop, we do body_fn in parallel (with a fixed number of threads M) and after that, selecting among ones which pass cond_fn. According to your experiment, 10e-7 failure rate (for M=8) is pretty small so we should definitely check it out. I'll mark this technique in my TODO list at #552.

**Comment by mattjj** [comment]
Thanks, @slinderman! My ulterior motive is to get you excited about working on JAX too ;)

**Comment by mattjj** [comment]
Thanks so much for this PR! Just a heads-up, we probably won't be able to review this until Wednesday.

**Comment by fehiepsi** [comment]
Thanks, please take your time @mattjj !

**Comment by mattjj** [comment]
(Sorry, being slow on this, but I haven't forgotten about it!)

**Comment by fehiepsi** [comment]
(@mattjj Thanks! No worries, this is just an enhancement. I understand that there are many other things which need to address first. ;) )

**Review by mattjj — APPROVED** [review]
Looks great to me! I learned about sampling gammas by reading this :)

**Comment by mattjj** [comment]
Sorry again for the long delay.

We asked the XLA team for some advice on how to make this kind of rejection sampler more performant, and an idea that came up is that we could generate gamma variates in a loop that consumes a smaller array, then scatter them out into a larger array until the larger array is full. That would reduce the amount of extra work done. But it's probably too tricky to pursue, and this approach is nice and simple!

The convention of calling into e.g. `lax.mul` and `lax.add` instead of using infix operators makes the code a bit harder to read, so we should consider switching that. For now this code is quite consistent with everything else in the file, though!

**Comment by mattjj** [comment]
Thanks again for this contribution! This is a totally new type of sampler for us, and probably opens the door for more to come. It's also useful for pushing forward our understanding of how to generate fast code with XLA for new kinds of computations.

**Comment by fehiepsi** [comment]
Thank you a lot @mattjj ! With this, I can add beta/dirichlet/studentT sampler in upcoming PRs. After that, I will address pathwise derivative for Gamma. :)

**Comment by PaulScemama** [comment]
I have a quick question:

To me, this Gamma distribution (sampler) seems to be parameterized by a single parameter `a`. I have only seen parameterizations of the Gamma distribution with *two* parameters -- the shape parameter and the rate parameter (for example see [here](https://en.wikipedia.org/wiki/Gamma_distribution)).

Can anyone explain this discrepancy? I'm sure there is a reason but I can't find one after scowering online for a bit. 

For example, how would I create a sample from a $Gamma(\alpha=1, \beta=3)$ distribution with the Jax sampler? 


Thank you!

**Comment by mattjj** [comment]
@PaulScemama thanks for the question!

I'm not sure but I would guess that it's because the rate or inverse scale parameter can be dropped without loss of generality, as the caller can rescale the sample by multiplication. That is, using the shape-rate parameterization if $X \sim \textrm{Gamma}(\alpha, \, 1)$ then $X / \beta \sim \textrm{Gamma}(\alpha, \, \beta)$. Or in code, we could say

```python
def gamma(key, alpha, beta, shape=None, dtype=dtypes.float_):
  return jax.random.gamma(key, alpha, shape, dtype) / beta
```

It's analogous to how you could drop the variance (or even the mean!) parameter from a Gaussian sampler function, since the caller can always scale (and/or add) to transform the distribution.

That said, it might arguably be a less confusing API to provide a rate or scale parameter. On the other hand, by not providing it, the user never has to figure out whether the API follows the rate parameterization or the scale parameterization! That's kind of nice...

WDYT?

**Comment by PaulScemama** [comment]
That definitely makes sense! Thank you for the thorough and clear answer. I agree that the user never has to figure out whether the API follows the rate/scale parameterization. However, I do think maybe the docs should say something like 

"we implement the standard gamma density. Dividing the sample by the rate is equivalent to sampling from $\text{Gamma}(\alpha, \text{rate})$. And multiplying the sample by the scale is equivalent to sampling from $\text{Gamma}(\alpha, \text{scale})$."

I'm not sure that wording is good 😅 but it could be useful. No worries if you think this is not needed or "overkill". Thanks again for answering the question!

**Comment by mattjj** [comment]
Good idea. Could you open a new issue so we don't forget to improve the docs that way?

**Comment by PaulScemama** [comment]
Definitely! Should I @ you in the issue?
